### DISCUSSION TITLE
Python -> AST string -> Python

### DISCUSSION THREAD

**Discussion by WiktorPrzetacznik** [astral-sh/ruff#13123] [Q&A]
Is there any easy way to convert Python code into AST, then store this AST as some readable format that could be converted back into a Python code?

**Comment by dhruvmanila** [comment]
Can you describe your use case? Are you looking for some kind of API or do you just want to use Ruff's Python parser? For the latter, you can do so in your Rust project (example, https://github.com/fpgmaas/deptry/blob/4fa5a91c8e8d3d14466921808181e479a7543859/Cargo.toml#L21-L22, https://github.com/fpgmaas/deptry/blob/4fa5a91c8e8d3d14466921808181e479a7543859/src/imports/shared.rs#L25-L28). And, we also have a `Generator` that can transform the AST into an equivalent Python string. There's more info in this discussion https://github.com/astral-sh/ruff/discussions/12500#discussioncomment-10144988.

The following code is a reference that you can use to perform this roundtrip:

https://github.com/astral-sh/ruff/blob/8db147c09d4a3c0621077858f3c6154bf8215461/crates/ruff_python_codegen/src/generator.rs#L1429-L1435

**Reply by WiktorPrzetacznik** [reply]
Thank you for your response. The problem is, I'm able to parse my sources to `Parsed<ModModule>`, but I'm looking for a way to somehow convert this `Parsed<ModModule>` to a string with an ability to convert it back.
Is it possible with current API?
As a workaround, I'm thinking of simply implementing serde on those structs in my version of the crate.

**Reply by dhruvmanila** [reply]
> The problem is, I'm able to parse my sources to `Parsed<ModModule>`, but I'm looking for a way to somehow convert this `Parsed<ModModule>` to a string with an ability to convert it back.

By "string", are you referring to the Python code equivalent to the parsed AST? If so, then the `Generator` does that. If not, then is it that you want to convert the AST structure to a raw string for which I think the debug representation would do but then there's no way to convert that string back to the structure. I think for that case adding serde should help. I'm curious to know the use-case for this.
